src/utils.py

```diff
@@ -8,86 +8,116 @@
+    def wrapped(charm: CharmBase) -> bool:
+        return bool(charm.model.relations[integration_name])
 
-    def decorator(func: CharmEventHandler) -> CharmEventHandler:
-        """A decorator, applied to any event hook handler, to block juju unit."""
+    return wrapped
 
-        @wraps(func)
-        def wrapper(charm: CharmBase, *args: EventBase, **kwargs: Any) -> Optional[Any]:
-            event, *_ = args
-            logger.debug(f"Handling event: {event}.")
 
-            for condition in conditions:
-                not_met, msg = condition(charm)
-                if not_met:
-                    charm.unit.status = BlockedStatus(msg)
-                    return None
+peer_integration_existence = integration_existence(PEER_INTEGRATION_NAME)
+kratos_integration_existence = integration_existence(KRATOS_INFO_INTEGRATION_NAME)
+hydra_integration_existence = integration_existence(HYDRA_ENDPOINTS_INTEGRATION_NAME)
+openfga_integration_existence = integration_existence(OPENFGA_INTEGRATION_NAME)
+ingress_integration_existence = integration_existence(INGRESS_INTEGRATION_NAME)
 
-            return func(charm, *args, **kwargs)
 
-        return wrapper  # type: ignore[return-value]
+def container_connectivity(charm: CharmBase) -> bool:
+    return charm.unit.get_container(WORKLOAD_CONTAINER).can_connect()
 
-    return decorator
 
+def certificate_existence(charm: CharmBase) -> bool:
+    oauth_integration_exists = charm.oauth_integration.is_ready()
+    cert_transfer_integration_exists = bool(
+        charm.model.relations[CERTIFICATE_TRANSFER_INTEGRATION_NAME]
+    )
 
-def wait_when(*conditions: Condition) -> Callable[[CharmEventHandler], CharmEventHandler]:
-    """A factory of decorators that put juju unit in wait status when a certain condition validation fails."""
+    return False if (oauth_integration_exists and not cert_transfer_integration_exists) else True
 
-    def decorator(func: CharmEventHandler) -> CharmEventHandler:
-        """A decorator, applied to any event hook handler, to put juju unit in wait status."""
 
-        @wraps(func)
-        def wrapper(charm: CharmBase, *args: EventBase, **kwargs: Any) -> Optional[Any]:
-            event, *_ = args
-            logger.debug(f"Handling event: {event}.")
+def openfga_store_readiness(charm: CharmBase) -> bool:
+    return charm.openfga_integration.is_store_ready()
 
-            for condition in conditions:
-                not_met, msg = condition(charm)
-                if not_met:
-                    event.defer()
-                    charm.unit.status = WaitingStatus(msg)
-                    return None
 
-            return func(charm, *args, **kwargs)
+def openfga_model_readiness(charm: CharmBase) -> bool:
+    return bool(charm.peer_data[charm._workload_service.version])
 
-        return wrapper  # type: ignore[return-value]
 
-    return decorator
+CONDITION_STATUS_REGISTRY = (
```

Comment:
This does not have to be in every charm. This is only because admin charm has a long list of condition checks.